source/memory_pool.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <assert.h>
+
+#include <aws/common/common.h>
+
+struct aws_memory_pool {
+    struct aws_allocator *alloc;
+    size_t arena_size;
+    size_t element_size;
+    uint8_t *arena;
+    void *free_list;
+    int overflow_count;
+};
+
+struct aws_memory_pool *aws_memory_pool_init(struct aws_allocator *alloc, size_t element_size, int element_count) {
+    size_t stride = element_size > sizeof(void *) ? element_size : sizeof(void *);
+    size_t arena_size = sizeof(struct aws_memory_pool) + stride * element_count;
+    struct aws_memory_pool *pool = (struct aws_memory_pool *)aws_mem_acquire(alloc, arena_size);
+
+    if (!pool) {
+        return NULL;
+    }
+
+    pool->alloc = alloc;
+    pool->arena_size = arena_size - sizeof(struct aws_memory_pool);
+    pool->element_size = element_size;
+    pool->arena = (uint8_t *)(pool + 1);
+    pool->free_list = pool->arena;
+    pool->overflow_count = 0;
+
+    /* Hook up singly linked list to represent the free list. */
+    for (int i = 0; i < element_count - 1; ++i) {
+        void **element_ptr = (void **)(pool->arena + stride * i);
+        void *next = (void *)(pool->arena + stride * (i + 1));
+        *element_ptr = next;
+    };
+
+    /* Last element points to NULL. */
+    void **last_element_ptr = (void **)(pool->arena + stride * (element_count - 1));
+    *last_element_ptr = NULL;
+
+    return pool;
+}
+
+void aws_memory_pool_clean_up(struct aws_memory_pool *pool) {
+    if (pool->overflow_count) {
+        aws_raise_error(AWS_ERROR_MEMORY_LEAK);
+        aws_mem_release(pool->alloc, pool);
+        assert(0);
+    } else {
+        aws_mem_release(pool->alloc, pool);
+    }
+}
+
+void *aws_memory_pool_try_acquire(struct aws_memory_pool *pool) {
+    if (pool->free_list) {
+        void *mem = pool->free_list;
+        pool->free_list = *((void **)pool->free_list);
+        return mem;
```

Comment:
Can we `memset(mem, 0, pool->element_size)` before returning this? I just hit a bug in my MQTT code because I was assuming that the memory the memory pool returns is 0'd.